sharkfuser/include/fusili/logging.h

```diff
@@ -52,6 +54,141 @@ struct [[nodiscard]] ErrorObject {
   bool operator!=(ErrorCode compareCode) const { return code != compareCode; }
 };
 
+// Utility function that returns true if an ErrorObject represents a successful
+// result.
+inline bool isOk(ErrorObject Result) { return Result.isOk(); }
+
+// Utility function that returns true if an ErrorObject represents an
+// unsuccessful result.
+inline bool isError(ErrorObject Result) { return Result.isFailure(); }
+
+// Utility function to generate an ErrorObject representing a successful result.
+inline ErrorObject ok() { return {ErrorCode::OK, ""}; }
+
+// Utility function to generate an ErrorObject representing an unsuccessful
+// result.
+template <typename S>
+  requires std::convertible_to<S, std::string>
+inline ErrorObject error(ErrorCode err, S &&errMsg) {
+  return ErrorObject(err, std::forward<S>(errMsg));
+}
+
+// Represents either an error or a value T.
+//
+// ErrorOr<T> represents the results of an operation. In the successful case it
+// provides pointer/optional<T> like semantics to the underlying T, on the
+// failure path it is convertible to an ErrorObject describing the failure
+// preventing T from being created.
+//
+// usage:
+//   ErrorOr<AST> buildAST() {
+//      ErrorOr<std::string> maybeBuffer = getBuffer();
+//      FUSILI_CHECK_ERROR(maybeBuffer);
+//      AST ast = buildAST(*maybeBuffer);
+//      return ok(ast);
+//   }
+template <typename T> class [[nodiscard]] ErrorOr {
+  using Storage = std::variant<T, ErrorObject>;
+
+  Storage storage;
+
+  // The intended consumption pattern is to use `isOk` and `isError` utility
+  // methods to check this class when converted to an ErrorObject.
+  bool has_value() const noexcept { return std::holds_alternative<T>(storage); }
+
+  // Friend declaration to allow ErrorOr<U> to access ErrorOr<T>'s private
+  // members
+  template <typename U> friend class ErrorOr;
+
+public:
+  // Construct successful case from anything T is constructable from.
+  template <typename U>
+    requires std::constructible_from<T, U &&>
+  ErrorOr(U &&val) : storage(std::in_place_type<T>, std::forward<U>(val)) {}
+
+  // Copy constructor, to allow for ErrorOr<const char *> ->
+  // ErrorOr<std::string> for example.
+  template <typename U>
+    requires std::is_constructible_v<T, U>
+  ErrorOr(const ErrorOr<U> &other) {
+    if (isOk(other)) {
+      storage = Storage(std::in_place_type<T>, *other);
+    } else {
+      storage = Storage(std::in_place_type<ErrorObject>,
+                        std::get<ErrorObject>(other.storage));
+    }
+  }
+
+  // Move constructor, to allow for ErrorOr<const char *> ->
+  // ErrorOr<std::string> for example.
+  template <typename U>
+    requires std::is_constructible_v<T, U>
+  ErrorOr(ErrorOr<U> &&other) {
+    if (isOk(other)) {
+      storage = Storage(std::in_place_type<T>, std::move(*other));
+    } else {
+      storage = Storage(std::in_place_type<ErrorObject>,
+                        std::move(std::get<ErrorObject>(other.storage)));
+    }
+  }
```

Comment:
If possible add tests to cover both cases (if not already done). In general recommend looking at code coverage reports for newly added source code. It should be auto-uploaded on the GHA for code coverage.